gisutils/algo.py

```diff
@@ -0,0 +1,43 @@
+from gisutils import raster
+
+def average_slope(df, dem, dem_aff):
+    """
+    Computes the average slope between the first and last coordinates of 
+    a line using the elevations from a reference Digital Elevation Map.
+
+    Output DataFrame has new column `average_slope` containing the result
+    of the slope calculation (float).
+
+    Parameters
+    ----------
+    df : pandas.DataFrame
+    dem : array
+    dem_aff : affine.Affine
+    
+    
+    Returns
+    -------
+    pandas.DataFrame
+    
+    """
+    
+    df = df.copy() #this may not be necessary since the function has it's own namespace
+    
+    for row in df.itertuples():
```

Comment:
rather than looping through the dataframe, it's offen more performant to use `df.apply(_helper_fxn, axis=1)`

basically you need to break the contents of this loop out into a helper function that takes a row + other args as input.